### DISCUSSION TITLE
Issues with std.traits

### DISCUSSION THREAD

**Discussion by ntrel** [dlang/PhobosV3-Design#30] [Ideas]
@LightBender mentioned this, see also @jmdavis's reply:
https://forum.dlang.org/post/mailman.726.1707039100.3719.digitalmars-d@puremagic.com

I'm posting this so people can reply with concrete improvements.

**Comment by rikkimax** [comment]
One thing to note: ``fullyQualifiedName`` shouldn't be brought over, it has been added to the compiler.

Other traits of similar complexity should be considered for implementation into the compiler.

Slightly unrelated, but involving getUDAs and getting of overloads https://issues.dlang.org/show_bug.cgi?id=23500

A lot of reflection stuff should be in the compiler: https://issues.dlang.org/show_bug.cgi?id=23559

**Reply by ntrel** [reply]
> fullyQualifiedName shouldn't be brought over, it has been added to the compiler

Thanks. It's undocumented, so I filed https://issues.dlang.org/show_bug.cgi?id=24472.

Likewise isNested, hasMember, classInstanceAlignment, isStaticArray, isAssociativeArray, isAbstractFunction, isFinalFunction, isAbstractClass, isFinalClass, isCopyable can probably be removed in favour of __traits.

**Reply by jmdavis** [reply]
> One thing to note: `fullyQualifiedName` shouldn't be brought over, it has been added to the compiler.

Thanks for pointing that out. I had not realized that that was a thing. And that's definitely the sort of trait where it's better for the compiler to do it if we can have that.

> Likewise isNested, hasMember, classInstanceAlignment, isStaticArray, isAssociativeArray, isAbstractFunction, isFinalFunction, isAbstractClass, isFinalClass, isCopyable can probably be removed in favour of __traits.

Some of those will need to be in phobos.sys.traits, because the `__traits` version allows too much. In particular, the `__traits` traits tend to treat enums as if they were their base type instead of as a separate type which implicitly converts to their base type like the type system as a whole does, which leads to bugs. std.traits is inconsistent about how it deals with such implicit conversions, and that's also caused problems. The current plan is for traits in phobos.sys.traits to give information about the exact types that they're given and not what those types implicitly convert to.

And some traits (e.g. `__traits(isIntegral, ...)`) get _really_ bad by allowing a whole hoard of types to the point that it's very difficult to know what they actually allow, making them pretty useless IMHO, whereas we really want a set of traits that are crystal clear about what they allow.

So, there are definitely cases where it's going to be appropriate to have a trait in phobos.sys.traits that does something similar to one in `__traits`, because the one in phobos.sys.traits is clearer and simpler (and we may or may not be able to make improvements to those `__traits` with editions later on). But yes, whereas std.traits often replaces stuff in `__traits` just so that you don't have to touch `__traits`, the current plan with phobos.sys.traits is to not include anything from `__traits` when there's a trait in `__traits` which already has that functionality and does the right thing. And it was agreed upon at a recent planning meeting that we wouldn't be trying to hide `__traits` like we did with std.traits.

Also, some traits will need to be in phobos.sys.traits simply so that they can be used as template predicates, since we don't have template lambdas, but at that point, the documentation should make it clear that the trait is there primarily so that it can be used as a template predicate and that it shouldn't normally be used when you don't need a template predicate - e.g. phobos.sys.traits currently has `isSomeSymbol` and `isSomeType`, because they're clearly needed for the most common use cases for phobos.sys.meta's `indexOf` (which is like std.meta's `staticIndexOf` except that it takes a template predicate for the comparison, thereby avoiding the mess that it is `staticIndexOf`s comparison and also allowing `indexOf` to be more generally useful). But with really simple traits like that, we should be limiting ourselves to the ones that are clearly going to be needed, and if template lambdas were a thing, then we could just skip such traits.

Because we need the traits and meta stuff in place to do much of anything else, I've slowly been doing PRs with the stuff that's most obviously needed (in part guided by what's needed for tests and documentation for the other traits and meta stuff, so the order has been a bit wacky), so some stuff is already in phobos.sys.traits and phobos.sys.meta, but there's definitely quite a ways to go, and it's good to hear about undocumented traits in `__traits` which remove the need for some traits in phobos.sys.traits - especially something as complex as `fullyQualifiedName`. And if anyone has concrete suggestions about an existing trait, have at it. In some cases, it's very clear what should be done with a trait when porting it over (or whether it should be ported over at all), whereas in others, it's far more debatable. So, I've just been working my way through std.traits and std.meta, doing the most obvious porting first. Some will likely be left behind (at least for now) simply because it's not entirely clear that they're needed.

LOL. One that I have no intention of porting over is `isCallable`. Even simply trying to clearly document what it currently does makes it clear that it's _way_ too complicated because of templates, property functions, and `opCall`. On the other hand, most code really doesn't care about whether a symbol is arbitrarily "callable". Rather, it cares about whether a particular expression will work - e.g. `foo(arg1, arg2)` - which goes beyond whether the symbol in question is callable in some generic sense. The main place that it matters is a trait like `functionAttributes`, and that could just check whether `__traits(getFunctionAttributes, ...)` works with the argument. And actually, that's another trait that we could consider not porting over because of `__traits`, though I think that we probably should, because having the enum of bit flags is useful and less error-prone (since you can screw up the strings when comparing against `__traits(getFunctionTraits, ...)` but will get errors if you screw up the enums). It also makes stuff like `SetFunctionAttributes` cleaner.

Regardless, I think that the name of the game here is to have traits that are clear and clearly useful and to not wrap `__traits` when wrapping it is purely to hide it. Too many of the traits in std.traits are not sufficiently clear (either because their documentation isn't good enough, or because they're too complex to be able to reasonably make clear even with good documentation), and some of them are clearly unnecessary (e.g. `isBoolean`). That being said, a large percentage of them can and should be ported over in one form or another, but they do need to be looked over one by one, and a number of them need tweaking.

**Reply by pbackus** [reply]
> Some of those will need to be in phobos.sys.traits, because the `__traits` version allows too much. In particular, the `__traits` traits tend to treat enums as if they were their base type instead of as a separate type which implicitly converts to their base type like the type system as a whole does, which leads to bugs. [...] The current plan is for traits in phobos.sys.traits to give information about the exact types that they're given and not what those types implicitly convert to.

IMO it would be a big mistake to fix this in Phobos without *also* fixing it in the language.

**Reply by jmdavis** [reply]
> IMO it would be a big mistake to fix this in Phobos without also fixing it in the language.

While I agree that it would be great to fix what `__traits` is doing, that's going to require Editions, because otherwise, it risks breaking existing code in subtle ways - and not in a way that we can use deprecations either. As it is, we've been stuck with problems in std.traits for years precisely because making any changes to traits breaks code in ways that we can't get around with deprecations. So, the fact that we're replacing std.traits gives us a golden opportunity to finally fix those traits where we couldn't before. We may or may not get a similar chance with `__traits` thanks to Editions.

So, we really can't assume that anything in `__traits` is going to be fixed when working on phobos.sys.traits. It's not going to be fixed right now, and it may or may not ever be fixed. Until/Unless `__traits` is fixed with Editions, we have to design Phobos v3 with what `__traits` currently does. And I see no reason to make Phobos do the wrong thing just because `__traits` does.

In the cases where a trait in phobos.sys.traits does something similar to a trait in `__traits`, the documentation explains the difference, so the situation is then clear (and in almost all cases, the trait in Phobos has a different name as well, reducing potential confusion). And if we are able to later fix the traits in `__traits`, all the better. I would love to see that happen. But we're throwing away this golden opportunity if we try to make phobos.sys.traits match `__traits`' behavior, just because we haven't been able to fix `__traits`.

**Reply by Herringway** [reply]
> LOL. One that I have no intention of porting over is `isCallable`. Even simply trying to clearly document what it currently does makes it clear that it's _way_ too complicated because of templates, property functions, and `opCall`. On the other hand, most code really doesn't care about whether a symbol is arbitrarily "callable". Rather, it cares about whether a particular expression will work - e.g. `foo(arg1, arg2)` - which goes beyond whether the symbol in question is callable in some generic sense. 

As much as I'd love a template like that, the best we can do with current language features would be painfully prone to false negatives from `__traits(compiles)`. Something like a [template matching trait](https://issues.dlang.org/show_bug.cgi?id=19448) is needed.

**Reply by jmdavis** [reply]
I'm specifically talking about using something like `__traits(compiles, foo(arg1, arg2))` or `is(typeof(foo(arg1, arg2)))` and not about adding a trait, because checking whether the call compiles is exactly what needs to be done if you need to check whether a particular function call will actually compile - especially when you're dealing with an arbitrary callable. I don't see how a trait for checking whether a template constraint would pass or not would even be relevant to the situation, since the callable doesn't even necessarily involve a template. Rather, this is about cases where a template constraint needs to check whether a particular callable is callable with a particular set of arguments, and that's very much a question of `__traits(compiles, ...)`. I don't see any other way to do that, particularly if you don't want to try to write a very complicated set of checks for lvalues, rvalues, type qualifiers, etc. where you're basically trying to duplicate the compiler's logic for function calls and inevitably are going to get it wrong because of how many factors are involved (and those factors can change as the language evolves, making such an attempt even more error-prone - e.g. they'll change when move constructors get added like they probably will be).

Of course, there are other situations where checking whether a template constraint passes without actually instantiating the template would be useful. For instance, that would be great for verifying that template constraints are well-written, since then you could use it to try to catch cases where the template constraint allows arguments that then fail to compile with the function body, since in most cases, that's not what you want, and it means that the template constraint needs to be fixed. But I really don't see how checking whether a particular template constraint passes or not really applies to this situation.

Regardless, my main point was that `std.traits.isCallable` really doesn't make any sense, because it's too complicated, and checking whether a particular symbol is callable in some arbitrary set of circumstances generally isn't useful. What almost all template constraints checking for callability care about is whether a particular symbol is callable with a particular set of arguments, which does inevitably take us to `__traits(compiles, ...)`, but that's a very common thing with traits and template constraints. There are cases where `__traits(...)` can be used to tell us what we need, and there are cases where we can compare types and infer stuff about them using `is` expressions, but there are quite a few cases where the only solution is `__traits(compiles, ...)` or `is(typeof(...)))`, much as it's good to encapsulate that in traits where we can to improve their usability and to reduce how often template constraints have to directly check for compilability. For instance, the traits for ranges are heavily based on whether particular pieces of code compile with the given template argument, and we're basically stuck doing that when checking for more complex stuff.

So, I'm not proposing any traits for checking for callability - quite the opposite really. I'm saying that `isCallable` doesn't make sense, and any traits that we do have related to whether something is callable need to be specific to the circumstances rather than checking for generic callability.

**Reply by Herringway** [reply]
> I'm specifically talking about using something like `__traits(compiles, foo(arg1, arg2))` or `is(typeof(foo(arg1, arg2)))` and not about adding a trait, because checking whether the call compiles is exactly what needs to be done if you need to check whether a particular function call will actually compile - especially when you're dealing with an arbitrary callable. I don't see how a trait for checking whether a template constraint would pass or not would even be relevant to the situation, since the callable doesn't even necessarily involve a template. Rather, this is about cases where a template constraint needs to check whether a particular callable is callable with a particular set of arguments, and that's very much a question of `__traits(compiles, ...)`. I don't see any other way to do that, particularly if you don't want to try to write a very complicated set of checks for lvalues, rvalues, type qualifiers, etc. where you're basically trying to duplicate the compiler's logic for function calls and inevitably are going to get it wrong because of how many factors are involved (and those factors can change as the language evolves, making such an attempt even more error-prone - e.g. they'll change when move constructors get added like they probably will be).
> 
> Of course, there are other situations where checking whether a template constraint passes without actually instantiating the template would be useful. For instance, that would be great for verifying that template constraints are well-written, since then you could use it to try to catch cases where the template constraint allows arguments that then fail to compile with the function body, since in most cases, that's not what you want, and it means that the template constraint needs to be fixed. But I really don't see how checking whether a particular template constraint passes or not really applies to this situation.
> 
> Regardless, my main point was that `std.traits.isCallable` really doesn't make any sense, because it's too complicated, and checking whether a particular symbol is callable in some arbitrary set of circumstances generally isn't useful. What almost all template constraints checking for callability care about is whether a particular symbol is callable with a particular set of arguments, which does inevitably take us to `__traits(compiles, ...)`, but that's a very common thing with traits and template constraints. There are cases where `__traits(...)` can be used to tell us what we need, and there are cases where we can compare types and infer stuff about them using `is` expressions, but there are quite a few cases where the only solution is `__traits(compiles, ...)` or `is(typeof(...)))`, much as it's good to encapsulate that in traits where we can to improve their usability and to reduce how often template constraints have to directly check for compilability. For instance, the traits for ranges are heavily based on whether particular pieces of code compile with the given template argument, and we're basically stuck doing that when checking for more complex stuff.
> 
> So, I'm not proposing any traits for checking for callability - quite the opposite really. I'm saying that `isCallable` doesn't make sense, and any traits that we do have related to whether something is callable need to be specific to the circumstances rather than checking for generic callability.

The problem with checking that something compiles is that it checks that something compiles. Error in the function body? false. Error in the parameters because you forgot to update the statement? false. Some other error? still just false! `__traits(compiles)` NEVER only tests exactly the thing you mean to test. ANY usage of `__traits(compiles)` or `is(typeof(...))` due to the lack of a precise alternative is a complete and utter failure of the language in my eyes.

The lack of this feature is why [one of my phobos PRs](https://github.com/dlang/phobos/pull/6172) has been stalled for years.

**Reply by jmdavis** [reply]
> The problem with checking that something compiles is that it checks that something compiles. Error in the function body? false. Error in the parameters because you forgot to update the statement? false. Some other error? still just false! __traits(compiles) NEVER only tests exactly the thing you mean to test. ANY usage of __traits(compiles) or is(typeof(...)) due to the lack of a precise alternative is a complete and utter failure of the language in my eyes.

Well, `__traits(compiles, ...)` tests exactly what it's meant to - whether a particular piece of code compiles. And of course, there are cases where a more precise alternative would be nice, but we have the language that we have, and even if we add a few more traits to `__traits` so that `__traits(compiles, ...)` is needed less, it'll still be needed in other situations, because that's basically a game of whack-a-mole, and realistically, there are cases where you really do need to check whether a particular piece of code compiles. And a lot of the time, that works perfectly fine.

Of course, there are situations where it becomes problematic - particularly when errors get eaten - and that can result in the wrong overload being used or a particular `static if` not being compiled in, but there's only so much that we can do about that. And arguably that is a failure of the language, but even then, being able to test a template constraint without instantiating a template would only help us in some of the cases. So, while it might be something that we should add to the language, it's just one improvement.

Ultimately, there's no way that we're going to have individual traits in `__traits` for everything, and `__traits(compiles, ...)` is going to need to be used pretty often. If we can reduce that with some additional traits in `__traits`, then arguably we should, but that's never going to cover everything. And there times where you genuinely want to test whether a particular piece of code compiles, because that's really what the template constraint needs to care about.

Given how template constraints and generic code work, ultimately, the solution is often to simply make sure that we test our code thoroughly (which we need to be doing anyway), and that catches a lot of these problems. Obviously, it's better when we can catch problems at compile time, but there are always going to be cases that we can't catch with compile-time checks - e.g. whether popFront actually pops a single element correctly.

So, while I can certainly see why you would consider it a failure of the language to use `__traits(compiles, ...)`, the reality of the matter is that we need to, and it's highly unlikely that that's ever going to change. We might get some `__traits` that make it so that we need to use `__traits(compiles, ...)` less, but it's not going away. There are way too many kinds of things that need to be tested for for the compiler to ever account for them all.
